audiovario.c

```diff
@@ -76,58 +77,62 @@ float change_volume(float delta){
 
-  phase= fmodf(phase,2*m_pi);
   if (phase <m_pi) return (phase-m_pi/2)*2/m_pi;
-  else return 1-(phase-m_pi)*2/m_pi;
+  return 1-(phase-m_pi)*2/m_pi;
 }
 
-void synthesise_vario(float val, int16_t* pcm_buffer, size_t frames_n, t_vario_config *vario_config){
+void synthesise_vario(float val, int16_t* pcm_buffer, size_t frames_n, t_vario_config *vario_config) {
 	unsigned int j;
 	float freq, pulse_freq; 
-
-   for(j=0;j<frames_n;j++) {
-     if (mute || (val > vario_config->deadband_low && val < vario_config->deadband_high)) pcm_buffer[j]=0;
-     else {
-       if (val > 0){
-         pulse_freq = (val > 0.5)? float(sample_rate)/(vario_config->pulse_length/(val*vario_config->pulse_length_gain)) : (float(sample_rate)/(float(vario_config->pulse_length*2)));
-         freq= vario_config->base_freq_pos+(val*vario_config->freq_gain_pos);
-         pcm_buffer[j]=pulse_syn( float(j)*m_pi*2.0/float(sample_rate)*pulse_freq+pulse_phase_ptr, vario_config->pulse_rise,vario_config->pulse_fall,vario_config->pulse_duty) * 327.67*volume*triangle(float(j)*m_pi*2.0/sample_rate*freq+phase_ptr);
-       }
-	   else{
-         freq= vario_config->base_freq_neg / (1.0-val*vario_config->freq_gain_neg);
-         pcm_buffer[j]=327.67*volume*triangle(float(j)*m_pi*2.0/float(sample_rate)*freq+phase_ptr);
-      }
-
-     }
-     synth_ptr++;
-   }
-   phase_ptr=float(j)*m_pi*2.0/float(sample_rate)*freq+phase_ptr;
-   phase_ptr= fmodf(phase_ptr,2*m_pi);
-   
-   pulse_phase_ptr=float(j)*m_pi*2.0/float(sample_rate)*pulse_freq+pulse_phase_ptr;
-   pulse_phase_ptr= fmodf(pulse_phase_ptr,2*m_pi);
-   
+	float freq_upd, pulse_freq_upd, temp;
+
+	if (mute || (val > vario_config->deadband_low && val < vario_config->deadband_high)) { 
+		memset(pcm_buffer,0,frames_n*sizeof(int16_t));
+	} else {
+		if (val>0) {
+			pulse_freq = (val > 0.5)? float(sample_rate)/(vario_config->pulse_length/(val*vario_config->pulse_length_gain)) : (float(sample_rate)/(float(vario_config->pulse_length*2)));
+         		freq= vario_config->base_freq_pos+(val*vario_config->freq_gain_pos);
+			temp=2.0*m_pi/float(sample_rate);
+			pulse_freq_upd=fmodf(temp*pulse_freq,2.0*m_pi);
+			freq_upd=fmodf(temp*freq,2.0*m_pi);
+			for (j=0;j<frames_n;++j) {
+				pcm_buffer[j]=(int)(round(pulse_syn(pulse_phase_ptr, vario_config->pulse_rise,vario_config->pulse_fall,vario_config->pulse_risei,vario_config->pulse_falli,vario_config->pulse_duty) * 327.67*volume*triangle(phase_ptr)));
+				pulse_phase_ptr+=pulse_freq_upd;
+				phase_ptr+=freq_upd;
+				if (pulse_phase_ptr>=2.0*m_pi) pulse_phase_ptr-=2.0*m_pi;
+				if (phase_ptr>=2.0*m_pi) phase_ptr-=2.0*m_pi;
+			}
+		} else {
+			freq= vario_config->base_freq_neg / (1.0-val*vario_config->freq_gain_neg);
+			freq_upd=fmodf(2.0*m_pi/float(sample_rate)*freq,2.0*m_pi);
+			for (j=0;j<frames_n;++j) {
+				pcm_buffer[j]=327.67*volume*triangle(phase_ptr);
+				phase_ptr+=freq_upd;
+				if (phase_ptr>=2.0*m_pi) phase_ptr-=2.0*m_pi;
+
+			}
+		}
+	}
+	synth_ptr+=frames_n;
```

Comment:
What is changed here?

Reply:
The main thing this does is rewrite the synthesise_audio function to improve speed.  It:

1) Moves the if val statements outside the loop because the value doesn't change.
2) moves the freq and pulse_freq assignments outside of the loop because they don't change.
3)old code has phase = freq/sample_rate*j +phase_ptr.  New code just keeps incrementing phase_ptr by a pre-calculated value.

All changes are geared to improving efficiency.  Doesn't really change functionality.  I put out the PR to show the changes I was talking about.  I did not see as much improvement as I had hoped.  But it did seem to work fine.  Will test more tommorow.

Oddly the old code assigns a float to an integer (buffer_pcm) this doesn't seem valid or at least doesn't seem like a good idea.